### ISSUE TITLE
API endpoints that uses multiple queries might result in terrible things

### ISSUE THREAD

**Issue by clarafu** [concourse/concourse#4373]
In the lidar drills environment, we found that if the API connection pool was maxed out in connections, the web memory usage increased at a crazy rate.

@vito theorizes that it is because for this specific case where the API requests were for ListAllJobs for the dashboard, it would result in three different db queries (from VisibleJobs call) which each acquire a connection (https://github.com/concourse/concourse/blob/master///atc/db/pipeline.go#L1045:20).

If there was a backlog of requests due to slowness with the db, most of the requests will wait to acquire a connection on each of these three calls. However instead of one request acquiring a connection, loading all the data, and returning, it will compete with future requests which *also* try to acquire a connection three times. While it's waiting for the second or third database call, it will be holding whatever data it has loaded so far, which leads to a memory leak.

A possible solution could be to have all API endpoints use only one transaction.

**Comment by clarafu** [comment]
The metric below shows the dramatic increase in web memory that was a result of the bug described. (The purple line with web/1)

<img width="550" alt="Screen Shot 2019-09-05 at 11 16 33 AM" src="https://user-images.githubusercontent.com/19290703/64355123-a6ca5000-cfce-11e9-8dc7-a2507d46a89a.png">

The heap dump diagram below is for lidar-drills when the web/1 node started to misbehave. 

![heap](https://user-images.githubusercontent.com/19290703/64354897-4509e600-cfce-11e9-8b86-bc0c473ceb7b.png)

As you can see from the diagram, the "reflect unsafe_NewArray" block was abnormally huge (using approx 1 GB of memory).

**Comment by xtremerui** [comment]
We have reproduced the issue in our test env and following is a summary.

===========================

Before:
Concourse 5.2.4 is deployed with 5 webs and 40 workers with 1500 pipelines, 15000 jobs. Then we make all pipelines public and use a script 
```
while true; do seq 1 1000 | xargs -P 1000 -n1 bash -c 'curl http://lidar-drills.concourse-ci.org/api/v1/jobs'; sleep 1; done
``` 
to curl /api/v1/jobs endpoint in order to satuate the DB connection pool for API.

Then we observe the behaviour described in this issue, where web node memory got shoot up while API pool is saturated by the above scrpit between 11:14-11:15. After we stopped the script the memory is still high and 3 of 5 web nodes crashed.
 
![Screen Shot 2019-11-15 at 10 31 28 AM](https://user-images.githubusercontent.com/1585949/68957001-2e82c680-0797-11ea-91d3-8cc3ecf94ef4.png)
![Screen Shot 2019-11-15 at 10 33 11 AM](https://user-images.githubusercontent.com/1585949/68957011-317db700-0797-11ea-9c62-1eab5f19de65.png)

In the heap dump, there is large memory allocated for `ListAllJobs` 
![profile001](https://user-images.githubusercontent.com/1585949/68960185-a6ec8600-079d-11ea-95db-3d7abdd0a39f.png)

============================

After:
The test env got upgraded to Concourse 5.2.5 with fix of #4761. Running the same script we observe the web node memory is stable even the API pool is saturated for a longer period of time.
![Screen Shot 2019-11-15 at 10 37 45 AM](https://user-images.githubusercontent.com/1585949/68960850-04350700-079f-11ea-8650-b711bc9bbd91.png)
![Screen Shot 2019-11-15 at 10 38 15 AM](https://user-images.githubusercontent.com/1585949/68960853-05feca80-079f-11ea-8c4f-237a1323278d.png)

 Meanwhile the size of heap dump dropped from 885kb to 12kb, and it doesn't have objects allocated to `ListAllJobs` anymore.
![profile002](https://user-images.githubusercontent.com/1585949/68961029-5aa24580-079f-11ea-801f-4ec78ff024c8.png)
